worker/src/queue.rs

```diff
@@ -1,124 +1,255 @@
+    pub fn raw_iter(&self) -> RawMessageIter {
+        let messages = self.inner.messages();
+        RawMessageIter {
+            range: 0..messages.length(),
+            array: messages,
         }
     }
+}
 
+impl<T: DeserializeOwned> MessageBatch<T> {
     /// An array of messages in the batch. Ordering of messages is not guaranteed.
-    pub fn messages(&self) -> Result<Vec<Message<T>>>
-    where
-        T: DeserializeOwned,
-    {
+    pub fn messages(&self) -> Result<Vec<Message<T>>> {
         self.iter().collect()
     }
+
+    /// Iterator for messages in the message batch. Ordering of messages is not guaranteed.
+    pub fn iter(&self) -> MessageIter<T> {
+        let messages = self.inner.messages();
+        MessageIter {
+            range: 0..messages.length(),
+            array: messages,
+            marker: PhantomData,
+        }
+    }
 }
 
-pub struct MessageIter<'a, T> {
-    range: std::ops::Range<u32>,
-    array: &'a Array,
-    timestamp_key: &'a JsValue,
-    body_key: &'a JsValue,
-    id_key: &'a JsValue,
-    data: PhantomData<T>,
+impl<T> From<MessageBatchSys> for MessageBatch<T> {
+    fn from(value: MessageBatchSys) -> Self {
+        Self {
+            inner: value,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// A message that is sent to a consumer Worker.
+pub struct Message<T> {
+    inner: MessageSys,
+    body: T,
+}
+
+impl<T> Message<T> {
+    /// The body of the message.
+    pub fn body(&self) -> &T {
+        &self.body
+    }
+
+    /// The body of the message.
+    pub fn into_body(self) -> T {
+        self.body
+    }
 }
 
-impl<T> MessageIter<'_, T>
+impl<T> TryFrom<RawMessage> for Message<T>
 where
     T: DeserializeOwned,
 {
-    fn parse_message(&self, message: &JsValue) -> Result<Message<T>> {
-        let js_date = js_sys::Date::from(js_sys::Reflect::get(message, self.timestamp_key)?);
-        let id = js_sys::Reflect::get(message, self.id_key)?
-            .as_string()
-            .ok_or(Error::JsError(
-                "Invalid message batch. Failed to get id from message.".to_string(),
-            ))?;
-        let body = serde_wasm_bindgen::from_value(js_sys::Reflect::get(message, self.body_key)?)?;
-
-        Ok(Message {
-            id,
+    type Error = Error;
+
+    fn try_from(value: RawMessage) -> std::result::Result<Self, Self::Error> {
+        let body = value.body()?;
+        Ok(Self {
+            inner: value.inner,
             body,
-            timestamp: Date::from(js_date),
         })
     }
 }
 
-impl<T> std::iter::Iterator for MessageIter<'_, T>
+/// A message that is sent to a consumer Worker.
+pub struct RawMessage {
+    inner: MessageSys,
+}
+
+impl RawMessage {
+    /// The body of the message.
+    pub fn body<T>(&self) -> Result<T>
+    where
+        T: DeserializeOwned,
+    {
+        Ok(serde_wasm_bindgen::from_value(self.raw_body())?)
```

Comment:
I think I would have expected this to return a `JsValue` and then not have the `Deserialize` trait bound. Then, when its converted to a `Message` above, it is deserialized.

Reply:
👍, done in https://github.com/cloudflare/workers-rs/pull/335/commits/09795506f4cefb4c9d69f95856f09ba2be4efe8d